### ISSUE TITLE
Build fails for ARM Debian on Android (Debian noroot)

### ISSUE THREAD

**Issue by Grima04** [ptitSeb/box86#118]
Hello,
I tried to build box86 on Debian for Android (The app is called Debian noroot on the PlayStore) but it fails when trying to build it with the message "error: unrecognized command line option '-marm' ; did you mean '-fasm'?".
I used the compile instructions you provided under the compile instructions "for Other ARM Linux".
My device is a Samsung Galaxy S7 with an Exynos 8890 chipset.

Thanks in advance,
Grima04

![Screenshot_20200518-115623](https://user-images.githubusercontent.com/22496380/82202167-ceb1d780-9901-11ea-9661-da63682a9314.png)

**Comment by ptitSeb** [comment]
I'm pretty sure you are on arm64 environnement, and box86 NEEDS 32bits. It cannot be built as a 64bits binary. That's why the `-marm` command is not recognised.

**Comment by Grima04** [comment]
Hello, sorry for the late reply but I was busy with some university exams lately.
I have now checked the environment and I am indeed on an arm64 environment. I could try to get it to work with a chroot since the Debian app doesn't seem to offer a 32 Bit environment or alternatively test it with Linux Deploy if the chroot doesn't work for some reason. 
Since the fault is clearly on my end though, I think that this issue can be safely closed now.
I have a last question though. Is Android support for box86 planned in the future or is it not possible due to how box86 works? It would be nice to have an alternative to ExaGear on Android, especially when box86 gets Wine support, since ExaGear is now dead and there is currently no solution to have fast x86 emulation with 3D acceleration on Android.

**Comment by ptitSeb** [comment]
box86 is 32bits, wit Android pushing 64bits, I'm unsure.
Now, Android version of box86 may be possible, but you'll need a large bunch of arm lib on the Android side, not sure all exist (like the X11 ones). Plus some wrapping I guess, for some functions.

Technically, it's possible I guess, but not on my roadmap for now, as I have too many things to do for now.

**Comment by moocow1452** [comment]
I'm willing to take a poke at it. Using a Pixel 3a on the Android 11 beta. Termux could at least get the Cmake files created, userLAnd had a bit more trouble but I believe they support 32bit compiling options. We could fumble our way through something.

**Comment by laytewon** [comment]
Maybe a 32 bit linux chroot on android would be a good middle ground till there is a working android version.
Would love to do lots of game testing on android 10.

**Comment by truboxl** [comment]
Haven't figured out how to use Termux clang to build yet (install ndk-multilib and `--target` only can only get so far)

But Anlinux Ubuntu armhf environment seems to work fine on my aarch64 device (other distro should work too)
Copy the instruction from Anlinux app but dont execute all of it.
Edit the downloaded ubuntu.sh to download and install ubuntu-rootfs-armhf.tar.xz (change `dpkg...` to `echo arm`)

and within ubuntu environment, install gcc g++ cmake python3 and you are all set

I also seen others suggest to use `-m32` but its not a good idea as you will end up building armv4 binaries as shown by:
`gcc -m32 -E -dM - < /dev/null | grep -i arch`

**Comment by moocow1452** [comment]
@truboxl I was able to run a build, but it gave me these errors.

```
/root/box86/src/wrapped/wrappedgio2.c: In function ‘my_g_simple_async_result_new_error’:
/root/box86/src/wrapped/wrappedgio2.c:502:5: warning: ignoring return value of ‘vasprintf’, declared with attribute warn_unused_result [-Wunused-result]
     vasprintf(&tmp, fmt, b);
     ^~~~~~~~~~~~~~~~~~~~~~~
/root/box86/src/wrapped/wrappedgio2.c: In function ‘my_g_simple_async_report_error_in_idle’:
/root/box86/src/wrapped/wrappedgio2.c:546:5: warning: ignoring return value of ‘vasprintf’, declared with attribute warn_unused_result [-Wunused-result]
     vasprintf(&tmp, fmt, b);
     ^~~~~~~~~~~~~~~~~~~~~~~
```

Also test #11 for ctest failed. Everything else was great.

**Comment by ptitSeb** [comment]
those are not error, but harmless warning. I'll silence them, but really, that's nothing.

test #11 works for me. Is it still failling for you? (all the multithread test sometimes fail, on rare occasion, for unknown reason, timming and race condition between the dynarec building the block and the other thread maybe?).

**Comment by moocow1452** [comment]
Latest pull, no more errors and 11 passes. Now all we have to do is figure out literally everything else! I'm going to see what can be done with xSDL, going to try and get FTL running when I have a minute.

**Comment by ptitSeb** [comment]
You'll need OpenGL also for FTL. Do have some GL or GLES hardware accelleration in the chroot?

**Comment by moocow1452** [comment]
After some looking around, graphical acceleration is a bit of a obstacle for running a Linux Chroot on Android, requiring at the very least root access. People have apparently had some luck with Sparkle, see meefik/linuxdeploy#746 , but even that is held back by only being software accelerated. Maybe the best way forward is to start looking into ndk and making compatibility layer similar to Wine, but that's probably within the scope of an entirely new project.  Anyone else have insight?

**Comment by laytewon** [comment]
I was thinking maybe using the method Exagear Strategies used on android. They have a small Ubuntu distribution with modified Wine that they allow others to change as well.

![Screenshot_20200702-165632](https://user-images.githubusercontent.com/8674135/86413063-ed87e580-bc85-11ea-8fc5-db53c0e64766.png)

Maybe we could change the Ubuntu distribution with a different Wine arch version like armhf and to run box86 with it out of the box?

Here is the ubuntu distribution if you wanna have a look:
https://drive.google.com/file/d/15IiJBC6lD2rs_BLQFmZsIkOy0EG0W8YY/view?usp=drivesdk

**Comment by moocow1452** [comment]
> I was thinking maybe using the method Exagear Strategies used on android. They have a small Ubuntu distribution with modified Wine that they allow others to change as well.
> 

The graphics issue is the hitch though, Exagear could run something like FTL, but that was only after running patched versions on a rooted system with the Windows compatibility, and the guides I can find are in Russian. [Tran Khanh Duy](https://youtu.be/eekMi71PiNY) has had some success with GL-Streaming solutions with Chroot, so we may want to look there next.

**Comment by laytewon** [comment]
> The graphics issue is the hitch though, Exagear could run something like FTL, but that was only after running patched versions on a rooted system with the Windows compatibility, and the guides I can find are in Russian. [Tran Khanh Duy](https://youtu.be/eekMi71PiNY) has had some success with GL-Streaming solutions with Chroot, so we may want to look there next.

Well, we can combine box86 with Swiftshader  for android. 
(https://swiftshader.googlesource.com/SwiftShader)
`SwiftShader is a high-performance CPU-based implementation of the Vulkan and OpenGL ES graphics APIs12. Its goal is to provide hardware independence for advanced 3D graphics.`
Also:
`The SwiftShader libraries act as drop-in replacements for graphics drivers.
On Windows, most applications can be made to use SwiftShader's DLLs by placing them in the same folder as the executable. 
On Linux, the LD_LIBRARY_PATH environment variable or -rpath linker option can be used to direct applications to search for shared libraries in the indicated directory first.`

Its already being attempted with user modified builds of Exagear with qemu i386 but its so slow. Box86 would blow that out of the water.

**Comment by moocow1452** [comment]
Also asked around and found this solution.

https://github.com/jeffboody/loax
https://github.com/jeffboody/loax-server

It's been a while since it was created and will likely need to be polished up, but it doesn't seem to require root.

**Comment by truboxl** [comment]
I think I am leaning towards solutions suggested by moocow1452 as they do use the devices GPU

Another one I am thinking of is to expose the device's Android EGL libs to proot-ed environment and build/use libhydris which can use EGL libs compiled using bionic libc with glibc equivalent from the environment
And pray everything got accelerated using the device driver and GPU from that point

This does pose an interesting question for box86 on Android. I think box86 currently focuses on GCC and glibc. Android mainly uses Clang and bionic libc.

Using clang will produce a lot of warnings, not sure ptitSeb will fix them, so I will use `-Wno-*` to hide those away in `CFLAGS`.

Android has some custom ELF stuff, C stuff and missing pthreads stuff like pthread_cancel(). Most Linux apps we want to run are compiled with glibc. A native Android box86 exec would mostly be useless as it can run extremely rare x86 Android binary not x86 Linux binary.

I think Proot fakes the calls for the glibc Linux environment so that it think it works on bionic environment as long as it's on a compatible architecture. Box86 probably need to do more than just handle x86 to Arm but also glibc to bionic for Android.

So unless we add a lot of stuff to box86 just for Android to use Linux apps, we probably stuck with using box86 in a virtual Linux environment on Android.

There's also Android 32bit ABI bugs (y2038!) so we probably have 18 years left unless Google want to fix them which they probably don't have to as processors will be 64bit only in the future.

**Comment by Heasterian** [comment]
Looks like It's possible to build and run box86 on aarch64 Android inside of armhf system mounted in Proot using Termux. To do this I used modified Anlinux script. Just don't expect hardware acceleration, setting up VNC is also a bit tricky. Some packages just can't run.

**Comment by FreddieOliveira** [comment]
Guys, was anyone able to actually run Steam in Android? I'm having some trouble as described [here](#249)

**Comment by ptitSeb** [comment]
box86 can be build on Android using Termux. I'm closing the ticket, feel free to re-open it if needed.
